downloader/s3.go

```diff
@@ -0,0 +1,114 @@
+package downloader
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	sss "github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/simplifi/looking-glass/config"
+	"os"
+	"path"
+)
+
+type s3 struct {
+	awsSession session.Session
+	awsBucket  string
+	awsPrefix  string
+}
+
+func newS3(config config.DownloaderConfig) (Downloader, error) {
+	err := validateConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	awsSess, err := createAwsSession(config.AwsKey, config.AwsSecret, config.AwsRegion)
+	if err != nil {
+		return nil, err
+	}
+	downloader := s3{
+		awsSession: *awsSess,
+		awsBucket:  config.AwsBucket,
+		awsPrefix:  config.AwsPrefix,
+	}
+
+	return &downloader, nil
+}
+
+func validateConfig(config config.DownloaderConfig) error {
+	if config.AwsKey == "" {
+		return fmt.Errorf("configuration value cannot be empty: AwsKey")
+	}
+	if config.AwsSecret == "" {
+		return fmt.Errorf("configuration value cannot be empty: AwsSecret")
+	}
+	if config.AwsRegion == "" {
+		return fmt.Errorf("configuration value cannot be empty: AwsSecret")
+	}
+	if config.AwsPrefix == "" {
+		return fmt.Errorf("configuration value cannot be empty: AwsKey")
+	}
+	if config.AwsBucket == "" {
+		return fmt.Errorf("configuration value cannot be empty: AwsKey")
```

Comment:
Should be AwsBucket.  Is there a way to stringify variable names in go?

Reply:
I haven't been able to find a reliable way :(

I didn't really like the way I did this either, just wasn't sure of another way to handle it.

Reply:
Updated this to maybe be a little cleaner: https://github.com/simplifi/looking-glass/pull/12/commits/0ad40dc653e9ed62405fdd8c0dd10f3024a63fb0

Thoughts?

Reply:
Much better -- at least this way that kind of error would stick out a lot more